core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala

```diff
@@ -188,10 +188,10 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
       // SPARK-35009: avoid creating multiple monitor threads for the same python worker
       // and task context
       if (PythonRunner.runningMonitorThreads.add(key)) {
-        new MonitorThread(SparkEnv.get, worker, context).start()
+        new MonitorThread(SparkEnv.get, worker, writerThread, context).start()
```

Comment:
Hmm, seems only one `MonitorThread` is created for same python worker in reusing case. Then we will be only able to monitor one writer thread?

Reply:
Hmm, good catch. The lifetime of the MonitorThread is tied to the lifetime of the Python worker, so this won't work when the worker is reused.

I think it would be easier to just create a new WriterMonitorThread that is tied to the lifetime of the writer thread, rather than trying to stuff the functionality into the MonitorThread.

I'll prototype this and try to craft a unit test that exercises the case you mentioned. It seems like we should be able to see the effect with something like `rdd.mapPartitions(lambda iterator: iterator.take(1)).coalesce(1)`. It might be tricky to actually reproduce the deadlock in that case though.

cc @attilapiros

Reply:
I haven't been able to repro this yet, but I did make the change to move to a separate WriterMonitorThread.